src/doom/r_main.c

```diff
@@ -699,13 +702,38 @@ void R_InitLightTables (void)
    // [crispy] smooth diminishing lighting
     if (crispy->smoothlight)
     {
-	LIGHTLEVELS = 32;
-	LIGHTSEGSHIFT = 3;
-	LIGHTBRIGHT = 2;
-	MAXLIGHTSCALE = 48;
-	LIGHTSCALESHIFT = 12;
-	MAXLIGHTZ = 1024;
-	LIGHTZSHIFT = 17;
+#ifndef CRISPY_TRUECOLOR
+		LIGHTLEVELS = 32;
+		LIGHTSEGSHIFT = 3;
+		LIGHTBRIGHT = 2;
+		MAXLIGHTSCALE = 48;
+		LIGHTSCALESHIFT = 12;
+		MAXLIGHTZ = 1024;
+		LIGHTZSHIFT = 17;
+#else
+		if (crispy->truecolor)
+		{
+		// [crispy] if in TrueColor mode, use smoothest diminished lighting
+		LIGHTLEVELS = 256;
+		LIGHTSEGSHIFT = 0;
+		LIGHTBRIGHT = 15;
+		MAXLIGHTSCALE = 376;
+		LIGHTSCALESHIFT = 9;
+		MAXLIGHTZ = 1024;
+		LIGHTZSHIFT = 17;
+		}
+		else
+		{
+		// [crispy] else, use paletted paletted approach
```

Comment:
> paletted paletted

I apologize, was literally working on a last breath, yesterday's marathon with light values exhausted me completely. But promised word should be kept. Will do all requested corrections today.

Reply:
No need to apologize! I am more than thankful that you guys keep on working on the ports and dig up new interesting projects each time. I don't even start writing any line of code when I'm tired, which is... about always nowadays. 😪

Reply:
> * Visplanes lighting (a.k.a. `zlight`) is not as smooth as walls lighting (a.k.a. `scalelight`), [screenshot](https://github.com/user-attachments/assets/dd118409-0ea9-46aa-9788-8c6c67c9c262). Not sure what to do, visplanes already have widest amount of possible lights `MAXLIGHTZ = 1024`. Changing `LIGHTZSHIFT` just leads to inaccuracies, but changing visplanes lighting formula is a no-go. Another thing is, despite of we are "3D" space, walls and floors are absolutely different entities.

Yes, that's an issue, though a minor one. I guess one has to compare how both arrays are calculated. For `zlight[][]` it's `level = startmap - (FixedDiv((ORIGWIDTH/2*FRACUNIT), (j+1)<<LIGHTZSHIFT)>>LIGHTSCALESHIFT)/DISTMAP;` whereas for `scalelight[][]` it's `level = startmap - j*NONWIDEWIDTH/(viewwidth_nonwide<<detailshift)/DISTMAP;`.

Reply:
> * Fake contrast is simply destroying diminishing logics in this case. It's a must-have de-facto for paletted render, especially for 1x resultion, but here... Just have a look at difference [with](https://github.com/user-attachments/assets/8cdac55f-55ec-4e1c-96c8-cac36e51ed33) and [with out](https://github.com/user-attachments/assets/4cd38c1d-2484-42b5-93f5-6e69f6453270) it.

Now that we have 15 levels of `LIGHTBRIGHT`, you could aadapt fake contrast much finer against the wall's actual angle. But that wouldn't help with the case in your screenshots, though. I guess we need to keep it as it is now.

Reply:
> I guess one has to compare how both arrays are calculated.

Definitely worth to look at generated values, there must be some potential. Ideally, if it will be possible to correct the formula without too much changes, conditions and macroses. I'll try to investigate, it is definitely worth to walk this extra mile.

> Now that we have 15 levels of LIGHTBRIGHT, you could aadapt fake contrast much finer against the wall's actual angle. 

Hmm.. Changing `1` to `0` in bit-shifting seems to work, but probably it could be much better. Need to think, if needed at all, better just follow your recommendation. For comparison: [before](https://github.com/user-attachments/assets/b7265c6d-78cc-434b-a733-ed87ddc7059a) and [after](https://github.com/user-attachments/assets/3762ed0e-93aa-402a-a058-cced806ee6f8).

Reply:
> Hmm.. Changing `1` to `0` in bit-shifting seems to work, but probably it could be much better. Need to think, if needed at all, better just follow your recommendation. For comparison: [before](https://github.com/user-attachments/assets/b7265c6d-78cc-434b-a733-ed87ddc7059a) and [after](https://github.com/user-attachments/assets/3762ed0e-93aa-402a-a058-cced806ee6f8).

Well, no. Changing bit-shifting to `0` in the formula in `p_setup.c` simply reverts back to normal fake-contrast. That is, it only applies to strictly vertical and horizontal lines. The smooth color grading comes from the increasing distance to the player. 

What this code does is to apply *half* of fake contrast to lines that are aligned in a certain angle interval around `45°`, i.e. half-way between strictly vertical and horizontal. What we *could do*, though, is to spread the available 15 levels of `LIGHTBRIGHT` across more intervals between `0°` and `90°`, but that'd involve some more math. :wink:

Reply:
> but that'd involve some more math

You know me, I'm certainly bad with maths and formulas, so barely will be able to write something useful. 😟

> Ideally, if it will be possible to correct the formula without too much changes, conditions and macroses.
Won't budge that easy. But if not changing formula, here's a small improvement:
```
	    MAXLIGHTZ = 1024*8;
	    LIGHTZSHIFT = 17-2;
```
Result is barely noticeable, but have a close look at the floor: [before](https://github.com/user-attachments/assets/589c2486-b080-4b1d-8005-a1891249ccbe) and [after](https://github.com/user-attachments/assets/727056f1-c5a7-49cd-9aac-001b2c941ecc). Looking down reveals difference far better: [before](https://github.com/user-attachments/assets/ca47b3a4-5583-484e-a44f-d80a0ed1ff1f) and [after](https://github.com/user-attachments/assets/3d3cb29f-b8e4-4875-abb9-d21296ffdf62). I have a bad feeling that it is maximum that can be done without changing original formula, but at least it's something.

Reply:
Was walking home after job, and screenshots with lowered camera makes me think.  I'll gladly explain, but it can't be explained just in a few words, so sorry for a long read.

Both of screenshots were made with disabled fake contrast, I need a cleaner view of how light levels are aligned, while fake contrast can be considered as a post-processing effect.

To be more precise, I was thinking about [this](https://github.com/user-attachments/assets/c0e96726-09d9-41c2-a866-ed9fe29ddd57) misalignment. In _ideal_ or simply real world, such misalignment shouldn't happen, light levels must be perfectly aligned and placed in harmony. But this is not our case, because the Mission is to make lighting smooth, not change it's model. You are probably remember Doom 95 and iconic view of Entryway with high resolution? It look different there, obliviously because of lighting code wasn't properly updated for higher resolutions. It was totally okay for that time, and even Doom Legacy have same issue, though I don't know, was it intentional or by design, but it's still there nowadays.

But we are seeking for accuracy. And most likely, if lighting will be shifted for perfect alignment, whole game world's lighting model will look... Different. So, I'm suggesting to consider vanilla lighting formulas as a sacred codes. Necessary changes were made to support higher resolutions, it is a must have, but even that changes keeping original look&feel.

So, to summarize, what should I do next:
* Fine-tune `MAXLIGHTZ` and `LIGHTZSHIFT` for more smoothing. Multiplier `8` will probably consume even more memory, and probably it doesn't needed to be that big.
* Fine-tune walls smoothing a bit, I'm still unhappy with that "1% inaccuracy". It should be a tiny correction in `MAXLIGHTSCALE`, just need to find a right value.

Reply:
Nearly perfect smoothing, but it cost about **12** megabytes of memory (20 in total for smoother implementation). 😢 It feels like it is showing more proper colors in extreme distance, the tunnel I'm looking at is about 4300 map units long.

```
MAXLIGHTZ = 1024*10;
LIGHTZSHIFT = 17-3;
```
Less smooth and less voracious costs 4,7 megabytes and still:
```
MAXLIGHTZ = 1024*4;
LIGHTZSHIFT = 17-2;
```

For comparison: [current approach](https://github.com/user-attachments/assets/cba1bde7-3268-4138-8d61-682518d9212a), [less smooth](https://github.com/user-attachments/assets/7d4182b2-1f50-45a5-9cd6-d8a33cfd3fc9), [more smooth](https://github.com/user-attachments/assets/3417c8a7-992b-4a0b-9d2c-b9fb3f61fb44). `-timedemo` results are same for all three approaches.. 
Which one you would like to have?

Reply:
Something is still odd. No matter which color you take in RGB color space, it will never take more than 256 steps to fade it to black, because then all the color channels will be zero. So, it makes no sense you'll have to allocate 10 MB of color tables to achieve a fade to black for a palette of RGB colors. I'll have a look at it later...